audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,337 @@
+        }
+        return 0;
+    }
+
+    void flush()
+    {
+        m_read_head = m_write_head = 0;
+    }
+};
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+gstreamer_file_source::gstreamer_file_source( const std::string&                   path,
+                                              audio::ifstream_info::container_type container,
+                                              size_t                               stream )
+: m_pipeline( nullptr, gst_object_unref )
+, m_ring_buffer( new RingBuffer() )
+{
+    init_gstreamer();
+    setup_source( path, container );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+gstreamer_file_source::~gstreamer_file_source()
+{
+    gst_element_set_state( m_pipeline.get(), GST_STATE_NULL );
+    wait_for_async_operation();
+    m_pipeline.reset();
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+GstState gstreamer_file_source::wait_for_async_operation()
+{
+    GstState current_state = GST_STATE_VOID_PENDING;
+    GstState pending_state = GST_STATE_VOID_PENDING;
+
+    while ( gst_element_get_state( m_pipeline.get(), &current_state, &pending_state, GST_MSECOND )
+            == GST_STATE_CHANGE_ASYNC )
+    {
+        g_main_context_iteration( nullptr, FALSE );
+    }
+
+    return current_state;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::init_gstreamer()
+{
+    static bool gstreamer_initialized = false;
+
+    if ( !std::exchange( gstreamer_initialized, true ) )
+    {
+        gst_init( 0, nullptr );
+    }
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::setup_source( const std::string& path, audio::ifstream_info::container_type container )
+{
+    GstElement* sink        = prepare_pipeline( path );
+    auto        sinkpad     = gst_element_get_static_pad( sink, "sink" );
+    auto        caps        = gst_pad_get_current_caps( sinkpad );
+    auto        caps_struct = gst_caps_get_structure( caps, 0 );
+    fill_format_info( caps_struct, container );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+GstElement* gstreamer_file_source::prepare_pipeline( const std::string& path )
+{
+    m_pipeline.reset( gst_pipeline_new( "pipeline" ) );
+
+    GstElement* source    = gst_element_factory_make( "filesrc", "source" );
+    GstElement* decodebin = gst_element_factory_make( "decodebin", "decoder" );
+    GstElement* queue     = gst_element_factory_make( "queue", "queue" );
+    GstElement* sink      = gst_element_factory_make( "appsink", "sink" );
```

Comment:
You probably want to set the `caps` property on appsink to the format that you expect to see. E.g. signed 16 bit integers in native endianness? Or 32 bit floating point?

Also putting a converter before the queue might be useful then.

Reply:
I think the main point of the ni-media lib is its converting power. From what I understand, they try to apply a lot of c++ template foo in order to eliminate function calls etc. So the lib itself is a super fast format converter. Adding gstreamers audioconvert/audioresample elements probably wouldn't be the way NI guys want it to be, would it?

Reply:
You should nonetheless restrict the caps to everything that ni can handle and not to more, and let GStreamer take care of everything else

Reply:
Yes @hhoegelo we want the original format if possible. We convert to client format on the fly in istream >> operator. We support every combination of 8 / 16 / 24 / 32 / 64 bit fixed_point (little / big endian, signed / unsigned) and 32 / 64 bit floating point ( big / little endian ) arithmetic.

Reply:
So the solution would be to set caps with all those on the appsink, and add an audioconvert which would then do nothing unless one of the other cases appears.